feat(greenhousectl): setup greenhouse dashboard (#845)
cloudoperators/greenhouse: dev-env/localenv/README.md

# Setting up development environment

This handy CLI tool will help you to setup your development environment in no time.

## Prerequisites

- [docker](https://docs.docker.com/get-docker/)
- [KinD](https://kind.sigs.k8s.io/docs/user/quick-start/)
- [kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/)

## Usage

> You can use `greenhousectl` either by downloading the latest binary

Reviewer: Why do I need `greenhousectl` if all commands are `make`?

Author: Eventually the goal was to make all these commands executable via `greenhousectl` without having to clone any github repositories. But the involved tool chain to do this exceeds the basic `docker` `kind` `kubectl` pre-requisites.

I will remove the `greenhousectl` section

from [here](https://github.com/cloudoperators/greenhouse/releases)

Or you can build it from source by running the following command: `make cli`

> [!NOTE]  
> The CLI binary will be available in the `bin` folder

## Setting up the development environment

There are multiple local development environment setup available for the Greenhouse project. You can choose the one that
fits your needs.

> `All commands will spin up KinD clusters and setup the necessary components`

Reviewer: Can you add a note on how to use the kubeconfig to access the central cluster?
Not all Users (think UI dev) might know their way around KinD.

Author: Done.

Normally `KinD` will automatically merge the `KinD` kubeconfigs in the default `~/.kube/config` so the current-context will always be the `kind-greenhouse-admin` context. 

But there might be a scenario where they haven't setup `~/.kube` so the note will be helpful 👍

### Develop controllers locally and run the webhook server in-cluster

```shell
make setup-controller-dev
```

> [!NOTE]
> set the environment variable `CONTROLLERS_ONLY=true` in your debugger configuration
> If no environment variable is set, the webhook server will error out due to the missing certs

### Develop Admission Webhook server locally

```shell
make setup-webhook-dev
```

> [!NOTE]
> set the environment variable `WEBHOOK_ONLY=true` in your debugger configuration
> If both the controllers and the webhook server needs to be run locally, do not set any environment variable

### Running Greenhouse Dashboard in-cluster

```shell
make setup-dashboard
```

> [!NOTE]
> You will need to port-forward the cors-proxy service and the dashboard service to access the dashboard
> Information on how to access the dashboard is displayed after the command is executed

> ### Develop Greenhouse Dashboard locally

Reviewer: Can we do this more generic?

Like
### Run Greenhouse Core for UI development
- spin up everything (`make setup`?) --> We don't care if we have a running Dashboard **AND** a local running node process, right?
- port forward the proxy with the SA token (we might need to think on how to "mock" different access levels in a follow up.
- Give hint to use this as the `apiEndpoint` property (or however it is called) to running frontends.
- Point to Juno Dashboard docs for running the dashboard locally.

Author: Done

```shell

> make setup

Author: `make setup-dashboard` supports both dual mode of local dashboard development and also to visualize plugins during local development of controllers and also for plugin developers.

The `make setup` is a full greenhouse installation and should not be used for controllers and webhook development hence it is referenced only for local plugin development and ui development. If put in first and people execute this and run the operator locally there will be two controllers fighting with each other causing infinite reconciliations.

Hence sequentially following  - 

1. for controller development `make setup-controllers-dev`
2. for webjook development `make setup-webhook-dev` (Or both 1 & 2)
3. Dashboard development `make setup`
4. Testing Plugins locally `PLUGIN_DIR=<absolute-path-to-charts-dir> make setup`

This was my thought process of why `make setup` is not highlighted as a big heading. 

@uwe-mayer let me know your thoughts on this.

> make setup

Reviewer: OK. I added a suggestion on the introduction to the file. To give an overview of the use cases with links to the respective sections. After that I don't mind the order...

```

- This will install the operator, cors-proxy, sample organization with an onboarded remote cluster
- Additionally, it also creates a `appProps.json` `ConfigMap` in the `greenhouse` namespace

> - You can now retrieve the generated `appProps.json` in-cluster by executing

Reviewer: On second thought: we might not even need that since we provide the how to run the dashboard locally. The dashboard instance in the Cluster can be (quite) statically configured. WDYT?

> - You can now retrieve the generated `appProps.json` in-cluster by executing

Author: for local development of dashboard you can ignore the dashboard installed in the cluster. You only need to port-forward the `cors-proxy` to local port `9090`

 The installed dashboard is only useful for controller developers and plugin developers for visualizing the CRs.

The installed dashboard is pretty static except for the fact that it needs this `appProps.json` to be generated - to skip OAuth and have a SA Token that has enough privileges to access the resources. 

If an OIDC token is needed for developing scenarios that need Identity token then a guide can be put on how to retrieve an Identity token from an IDP (Postman / Insomnia) and replace it in `appProps.json` 
But it will not work locally because when this Identity token is forwarded to `KinD APIServer` it will reject it as unAuthorized.

> - You can now retrieve the generated `appProps.json` in-cluster by executing

Reviewer: I would remove (sorry can't suggest the whole idea)

```
You can now retrieve the generated appProps.json in-cluster by executing kubectl get cm greenhouse-dashboard-app-props -n greenhouse -o=json | jq -r '.data.["appProps.json"]'
Optionally you can also redirect this output to appProps.json in [Juno Repository](https://github.com/cloudoperators/juno/blob/main/apps/greenhouse/README.md)
```
As this seems to be very edge-casy. And might come into a dedicated doc page for the specific use case.
I would think in the majority of cases we do not need to temper with the in-cluster dashboard... WDYT?

  `kc get cm greenhouse-dashboard-app-props -n greenhouse -o=json | jq -r '.data.["appProps.json"]'`
- Optionally you can also redirect this output to `appProps.json`
  in [Juno Repository](https://github.com/cloudoperators/juno/tree/main/apps/greenhouse)
- Follow the instructions in the terminal to `port-forward` the cors-proxy service (ignore the `port-forward` of
  dashboard service)

> - Start the dashboard locally

Author: done

- `PluginDefinition(s)` can be applied

>   from [Greenhouse Extensions](https://github.com/cloudoperators/greenhouse-extensions) repository

Reviewer: So you provide the GH extensions helm charts? Nice!

Can we provide a sample `kubectl` command, e.g.:

```bash
kubectl --kubeconfig=<path-to-kind-kubeconfig> apply -f https://github.com/cloudoperators/greenhouse-extensions/blob/main/exposed-services/plugindefinition.yaml
```

Author: I will delete this section as there is a PR on the [greenhouse extensions](https://github.com/cloudoperators/greenhouse-extensions/pull/579) that will provide a `Kustomize` to pull these resources remotely.

### Test Plugin / Greenhouse Extension charts locally

```shell
PLUGIN_DIR=<absolute-path-to-charts-dir> make setup
```

- This will install a full running setup of operator, dashboard, sample organization with an onboarded remote cluster

> - Additionally, it will mount the plugin charts directory on to the `node` of the `KinD` cluster

Reviewer: ```suggestion
- Additionally, it will mount the plugin charts directory on to the `node` of the `KinD` cluster. Make sure to point `PLUGIN_DIR` to the directory containing the `Chart.yaml` of the `Plugin` you want to try out.
```

Author: The `PLUGIN_DIR` is an env where one can specify any base folder that can contain one or multiple charts. example mounting the entire repository of `greenhouse-extensions`.

In the plugin definition `.spec.helmChart.name` should always point to 'local/plugins` + `/<base-dir-with-chart-yaml>` 

The PR https://github.com/cloudoperators/greenhouse-extensions/pull/579 also has a script that generates these local plugin definitions for the developers.

Reviewer: I added a second comment to explicitly mention the `local/plugins` path...

- The operator deployment has a hostPath volume mount to the plugin charts directory from the `node` of the `KinD`
  cluster

> You would need to apply the `PluginDefinition(s)` of the chart that needs to be tested.

Reviewer: Can we make this a little more simple and hopefully concrete? Can you swap the `<path-to-chart` in the yaml with the correct path within the hosted volume in the KinD? Same as with the `<your-kind-config>` part.

```suggestion
To test your local Chart (now mounted to the KinD cluster) with a `plugindefinition.yaml` you would need to adjust `PluginDefinition.Spec.helmChart` to use the local chart.
E.g.
```yaml

You would need to apply the `plugindefinition.yaml` of the chart that needs to be tested.

```yaml

apiVersion: greenhouse.sap/v1alpha1
kind: PluginDefinition
metadata:
  name: <your-plugin-name>
spec:
  ...
  helmChart:
    name: <path-to-chart>
    repository: ''
    version: ''
...
```

Apply this `PluginDefinition` by
```bash
kubectl --kubeconfig=<your-kind-config> apply -f plugindefinition.yaml
```

Author: done

Write the webhook certs to `/tmp/k8s-webhook-server/serving-certs`

> [!NOTE]
> The deployed controller-manager is running in webhook only mode, but it does not receive any requests

Now you can run the webhook server and the controllers locally

Since both need to be run locally no `CONTROLLERS_ONLY` or `WEBHOOK_ONLY` environment variables are needed in your
debugger configuration

> [!NOTE]
> The dev setup will modify the webhook configurations to have 30s timeout for the webhook requests, but
> when break points are used to debug webhook requests, it can result into timeouts.
> In such cases, modify the CR with a dummy annotation to re-trigger the webhook request and reconciliation

### Running Greenhouse Dashboard in-cluster

```shell
make setup-dashboard
```

> [!NOTE]
> You will need to port-forward the cors-proxy service and the dashboard service to access the dashboard
>
> Information on how to access the dashboard is displayed after the command is executed

### Run Greenhouse Core for UI development

```shell
make setup
```

- This will install the operator, cors-proxy, sample organization with an onboarded remote cluster

> - Additionally, it also creates a `appProps.json` `ConfigMap` in the `greenhouse` namespace
> - You can now retrieve the generated `appProps.json` in-cluster by executing
>   `kubectl get cm greenhouse-dashboard-app-props -n greenhouse -o=json | jq -r '.data.["appProps.json"]'`
> - Optionally you can also redirect this output to `appProps.json`
>   in [Juno Repository](https://github.com/cloudoperators/juno/blob/main/apps/greenhouse/README.md)
> - Follow the instructions in the terminal to `port-forward` the cors-proxy service (ignore the `port-forward` of
>   dashboard service)
> - After port-forwarding `cors-proxy` service, it should be used as `apiEndpoint` in `appProps.json`
> - Start the dashboard locally (more information on how to run the dashboard locally can be found in
>   the [Juno Repository](https://github.com/cloudoperators/juno/blob/main/apps/greenhouse/README.md)

Reviewer: ```suggestion
- port-forward the `cors-proxy` by `kubectl port-forward svc/greenhouse-cors-proxy 9090:80 -n greenhouse &`
- port-forward the `dashboard` by `kubectl port-forward svc/greenhouse-dashboard 5001:80 -n greenhouse &`
- Access the local `demo` organization on the Greenhouse dashboard on [localhost:5001](http://localhost:5001/?org=demo)

```

> ### Test Plugin / Greenhouse Extension charts locally

Reviewer: ```suggestion
### ### Run Greenhouse Core for UI development
- Startup the environment as in [Run local Greenhouse](#run-local-greenhouse)
- An `appProps.json` `ConfigMap` is created  in the `greenhouse` namespace to configure the dashboard.
- You can now retrieve the generated `appProps.json` in-cluster by executing
  `kubectl get cm greenhouse-dashboard-app-props -n greenhouse -o=json | jq -r '.data.["appProps.json"]'`
- Optionally you can also redirect this output to `appProps.json`
  in [Juno Repository](https://github.com/cloudoperators/juno/blob/main/apps/greenhouse/README.md)
- After port-forwarding `cors-proxy` service, it should be used as `apiEndpoint` in `appProps.json`
- Start the dashboard locally (more information on how to run the dashboard locally can be found in
  the [Juno Repository](https://github.com/cloudoperators/juno/blob/main/apps/greenhouse/README.md)
### Test Plugin / Greenhouse Extension charts locally
```

> # Setting up development environment

Reviewer: ```suggestion
# Setting up a local development environment

Greenhouse provides a couple of cli commands based on `make` to run a local Greenhouse instance.
This local instance is intended to support
- trying out Greenhouse locally
- developing Greenhouse core functionality:
    - controllers
    - webhooks
- developing Greenhouse `Plugins`
- developing UIs interacting with the Greenhouse API (kubernetes apiserver) such as:
    - Greenhouse dashboard UIs
    - `Plugin` Uis
```

Maybe we can link to the respective sections?

  `kubectl get cm greenhouse-dashboard-app-props -n greenhouse -o=json | jq -r '.data.["appProps.json"]'`
- Optionally you can also redirect this output to `appProps.json`
  in [Juno Repository](https://github.com/cloudoperators/juno/blob/main/apps/greenhouse/README.md)
- Follow the instructions in the terminal to `port-forward` the cors-proxy service (ignore the `port-forward` of
  dashboard service)
- After port-forwarding `cors-proxy` service, it should be used as `apiEndpoint` in `appProps.json`
- Start the dashboard locally (more information on how to run the dashboard locally can be found in
  the [Juno Repository](https://github.com/cloudoperators/juno/blob/main/apps/greenhouse/README.md)

### Test Plugin / Greenhouse Extension charts locally

```shell
PLUGIN_DIR=<absolute-path-to-charts-dir> make setup
```

- This will install a full running setup of operator, dashboard, sample organization with an onboarded remote cluster
- Additionally, it will mount the plugin charts directory on to the `node` of the `KinD` cluster
- The operator deployment has a hostPath volume mount to the plugin charts directory from the `node` of the `KinD`
  cluster

> To test your local Chart (now mounted to the KinD cluster) with a `plugindefinition.yaml` you would need to adjust `.spec.helmChart.name` to use the local chart.

Reviewer: ```suggestion
To test your local Chart (now mounted to the KinD cluster) with a `plugindefinition.yaml` you would need to adjust `.spec.helmChart.name` to use the local chart. With the provided mounting mechanism it will always live in `local/plugins/` within the KinD cluster.
```
